clang/lib/Sema/SemaExpr.cpp

```diff
@@ -218,14 +218,24 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
       bool IsRuntimeEvaluated =
           ExprEvalContexts.empty() ||
           (!isUnevaluatedContext() && !isConstantEvaluated());
+      bool IsEsimdPrivateGlobal = isSYCLEsimdPrivateGlobal(VD);
       if (IsRuntimeEvaluated && !IsConst && VD->getStorageClass() == SC_Static)
         SYCLDiagIfDeviceCode(*Locs.begin(), diag::err_sycl_restrict)
             << Sema::KernelNonConstStaticDataVariable;
       // Non-const globals are allowed for SYCL explicit SIMD.
-      else if (IsRuntimeEvaluated && !isSYCLEsimdPrivateGlobal(VD) &&
-               !IsConst && VD->hasGlobalStorage() && !isa<ParmVarDecl>(VD))
+      else if (IsRuntimeEvaluated && !IsEsimdPrivateGlobal && !IsConst &&
+               VD->hasGlobalStorage())
         SYCLDiagIfDeviceCode(*Locs.begin(), diag::err_sycl_restrict)
             << Sema::KernelGlobalVariable;
+      // ESIMD globals cannot be used in a SYCL context.
+      else if (IsRuntimeEvaluated && IsEsimdPrivateGlobal &&
+               VD->hasGlobalStorage()) {
+        auto *FD = getCurFunctionDecl();
+        if (FD && !FD->hasAttr<SYCLSimdAttr>())
```

Comment:
What is the purpose of this?  If this is to detect SIMD vs not-SIMD mode, than this shouldn't be necessary.  The whole point of the deferred-diagnostic changes are to handle something like that.

Reply:
This is to prevent adding a deferred diagnostic if we already know that it won't be emitted.

Reply:
The check on 3419 of SemaSYCL is supposed to do that.  If it doesn't, that needs debugging.

Reply:
No, AFAIU, it isn't supposed to do that. It only determines whether the diagnostic will be immediate or deferred. In my case, I already know that it will be neither immediate nor deferred.

Reply:
I believe you're misunderstanding.  In no place where we diagnose should we need to check the mode we're in, thats the entire purpose of the EmissionReason parts.

So unless I'm missunderstanding the purpose of the SYCLSimdAttr (I thought it meant it was an esimd function?), we should be counting on the check in the SemaSYCL to suppress this.

Reply:
Yes, SYCLSimdAttr means it is an ESIMD function.

Consider 3 cases for this diagnostic:
1. ESIMD function -> should be skipped since ESIMD globals are allowed in it. `getEmissionReason(FD) & Reason` is `Esimd & Sycl`, so `K_Deferred`.
2. SYCL function -> ESIMD globals are NOT allowed in it; should be immediately diagnosed. `getEmissionReason(FD) & Reason` is `Sycl & Sycl`, so `K_ImmediateWithCallStack`.
3. Unknown function (don't know the context) -> should be deferred. `getEmissionReason(FD) & Reason` is `All & Sycl`, so `K_Deferred`.

The code you mentioned can't distinguish 1 from 3, which I do explicitly.

Reply:
Then we need to figure out a way to tell the difference. The whole purpose of the patch is to avoid having to distinguish this stuff on a per-diagnostic basis.

Reply:
Oh, actually, for the 'immediate' case, 1 and 3 are fine being the same.  It is up to the deferred diagnostics to check it later to tell the difference.

Reply:
Yes, in case 1 the diagnostic could be deferred and thrown out later. But I think it's better not to defer it if we know it already.

We could implement something generic, like: if we know the function context and the diagnostic reason (it is more specific than `All`), and they are different, then we could skip building a diagnostic. However, I don't think `SYCLDiagIfDeviceCode` interface allows to skip creating a diagnostic, so it's not a good place for such a generic check.

Reply:
I disagree, the whole purpose of this patch is to defer it and throw it out later.  Storing the diagnostic isn't expensive.

Reply:
> I disagree, the whole purpose of this patch is to defer it and throw it out later. Storing the diagnostic isn't expensive.

I believe that's not the purpose for the cases when we know the diagnostic will not be emitted anyway. 

Also, IMO, explicit check makes it more intuitive for people not familiar with the deferred diagnostic mechanism. They can see that the diagnostic is only emitted for the SYCL code. Without this explicit check, users could be confused: "When/Where is the function context gets checked?"

Anyway, I don't plan to die climbing that hill, if those arguments don't convince you, I agree to remove this check.

Reply:
I'm unconvinced.  The problem with the explicit checking is it adds a need to add that on all of our diagnostics, which is against the purpose of the way we've designed the deferred diagnostics system.  That ends up being a ton of extra mental overhead on a pure diagnostic basis.

The answer to "When/where is the function context gets checked" is in the 2 places we do it now, either in `SYCLDiagIfDeviceCode`, or in the `DeferredDiagnosticsEmitter`.

Reply:
Though... this makes me wonder if we need to split up getEmissionReason.  There is a slight subtly that I think the implementation of isn't quite right then.  

In the immediate diagnostic case, we actually care about (for a reason):
"do we KNOW this is/willbe emitted for an incompatible reason" vs "do we not know whether this would be emitted".
In the former case, we can ignore it.  In the latter, we need to defer it.  Currently, the former is represented properly, but the latter is represented by 'all'.  In reality, I suspect that needs to be 'none', and that function might need to differentiate between the two cases.

This makes me think that the current implementation of getEmissionReason is insufficient.  I might need time to think about where you should go from there.

Reply:
> The problem with the explicit checking is it adds a need to add that on all of our diagnostics

Actually, only for diagnostics for coexisting languages. But personally, I think this check makes sense for 2 reasons: 1) not doing unnecessary work and 2) code readability.

Reply:
I disagree on the readability front.  As for 'only for coexisting languages', that is both REALLY error prone and is a giant maintenance headache when adding new languages.  For example, you already forgot openmp.